sodetlib/operations/bias_dets.py

```diff
@@ -201,3 +201,408 @@ def bias_to_rfrac(S, cfg, rfrac=0.5, bias_groups=None, iva=None,
+        time.sleep(sleep_time)
+        safe_dc_biases = previous_dc_biases.copy()
+        for replace_bg in bg_overbias_needed:
+            try:
+                safe_dc_biases[replace_bg] = cfg.dev.bias_groups[replace_bg]["testbed_100mK_bias_voltage"]
+            except KeyError:
+                mask_bg = np.where(bsa_0.bgmap==replace_bg)[0]
+                v_norm_bl = np.nanmedian(iva.v_bias[iva.idxs[[mask_bg], 1]])
+                safe_dc_biases[replace_bg] = v_norm_bl
+                
+            
+        S.set_tes_bias_bipolar_array(safe_dc_biases)  
+        bsa_0 = bias_steps.take_bias_steps(S, cfg)
+        if make_plots:
+            fig, ax = bias_steps.plot_Rfrac(bsa_0)
+            fname = sdl.make_filename(S, 'post_overbiasing_Rfrac.png', plot=True)
+            fig.savefig(fname)
+            S.pub.register_file(fname, 'biasstep_rfrac', format='png', plot=True)
+            if not show_plots:
+                plt.close(fig)
+        
+        ## add in check if detectors are normal now
+        percentage_rn_0 = bsa_0.R0/bsa_0.R_n_IV
+        drop_from_normal = False
+        for bl in bg_overbias_needed:
+            mask_bg = np.where(bsa_0.bgmap==bl)[0]
+            per_bl_percentage_rn_0 = percentage_rn_0[mask_bg]
+            ## mask for normal
+            mask_normal = np.where((per_bl_percentage_rn_0 > 0.9) | (per_bl_percentage_rn_0 < -0.9))
+            ## if more than half of the detectors are normal
+            if len(mask_normal[0]) > 0.5*len(mask_bg):
+                bg_detectors_normal.append(bl)   
+                drop_from_normal = True
+
+## drop normal detector into transition first
+    while drop_from_normal:
+        S.log(f"some biaslines are still in normal state: {bg_detectors_normal}")
+        percentage_rn_0 = bsa_0.R0/bsa_0.R_n_IV
+        S.log("droping detectors from normal to transition")
+        previous_dc_biases_dfn = S.get_tes_bias_bipolar_array()
+
+        S.log(f"current tes bias voltages are: {previous_dc_biases_dfn}")
+        for bl in bg_detectors_normal:
+            mask_bg = np.where(bsa_0.bgmap==bl)[0]
+            v_sc = iva.v_bias[iva.idxs[[mask_bg], 0]]
+            v_norm = iva.v_bias[iva.idxs[[mask_bg], 1]]
+            v_spread = np.nanmedian(v_norm) - np.nanmedian(v_sc)
+            if previous_dc_biases_dfn[bl] > v_spread:
+                previous_dc_biases_dfn[bl] =  previous_dc_biases_dfn[bl] - 0.5 * v_spread
+            else:
+                previous_dc_biases_dfn[bl] =  previous_dc_biases_dfn[bl] - 0.3 * v_spread 
+                bg_detectors_normal.remove(bl)
+                S.log(f'biasline {bl} is approching 0 voltage, a confirmation new IV curve is recommended')
+        
+        S.set_tes_bias_bipolar_array(previous_dc_biases_dfn)
+        S.log(f"applying: {previous_dc_biases_dfn}")
+        S.log(f"waiting 10s for dissipation of UFM local heating")
+        time.sleep(10)
+        bsa_0 = bias_steps.take_bias_steps(S, cfg)
```

Comment:
If you do not need to run on all bias lines, you can specify the `bgs` arg here to just run on a selection of them. This may speed up the function.

Reply:
this is a good idea, but I think this should be added into the next generation of this code. This is because I currently rely on bg map for all 12 biaslines in this version of code.